FileSystem Tab: Inline renaming of files and folders

Wiwip opened the pull request:
The current SceneTree allows for direct inline edit of node names while the FileSystem docker does not. In fact, it still uses a popup text dialog. The pull request is a simple proposal to enable inline editing in the FileSystem docker while maintaining feature parity.

Effectively, it replaces the first image by the second.
![image](https://user-images.githubusercontent.com/5385314/204096084-f45a7845-c154-43c4-ac10-0570f6572082.png)

![image](https://user-images.githubusercontent.com/5385314/204096135-9916c883-84a6-4e62-85b5-7f48a1677894.png)

The addition of the `void set_editor_selection(int start, int end);` is maybe not the definite best solution, however there are no other means of accessing the LineEdit in thew Tree class (other than making it public or adding a getter). Additionally, I'm not sure why TreeItems have the editable property set to "not".

This is my first pull request for Godot so please let me know if there's anything I missed. This seemed like a rather simple change.

*Bugsquad edit:* Closes https://github.com/godotengine/godot-proposals/issues/4933

winston-yallow committed:
Add editor setting for spin slider sensibility

Jayman2000 committed:
Streamline and centralize bug reporting guides

Before this change, there was three different bug reporting guides:

- [“Filing an issue on GitHub”][1]
- [“Reporting bugs”][2]
- [The issue template][3]

This commit:

1. makes sure that [the issue template][3] contains all of the same
   information that [“Filing an issue on GitHub”][1] and
   [“Reporting bugs”][2] did and
2. makes [“Reporting bugs”][2] simply tell users to fill out the
   template.

The goal of this change is to make reporting bugs easier. This change
accomplishes that goal by presenting bug reporters with all of the
information they need to know on the bug reporting page itself.

This commit partially implements this proposal:
<https://github.com/godotengine/godot-proposals/discussions/4083>

[1]: https://docs.godotengine.org/en/stable/community/contributing/ways_to_contribute.html#filing-an-issue-on-github
[2]: ./CONTRIBUTING.md#reporting-bugs
[3]: ./.github/ISSUE_TEMPLATE/bug_report.yml

KoBeWi committed:
Prompt to confirm anim track delete on node delete

MewPurPur committed:
Light3D show scaling warning immediately

Lasuch69 committed:
Add NoiseTexture3D

KoBeWi committed:
Add center_grabber property to Slider

Chaosus commented:
Thanks for your contribution. You need to squash the commits as required by our pipeline (see https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html).

Wiwip commented:
Just squashed the commits as required.

Wiwip commented:
Fixed some style errors.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1562,7 +1562,12 @@ void FileSystemDock::_folder_removed(String p_folder) {
 
 void FileSystemDock::_rename_operation_confirm() {
-	String new_name = rename_dialog_text->get_text().strip_edges();
+	if (!tree->is_anything_selected()) {
+		return;
+	}
```
I think this condition is not necessary (`is_anything_selected()` is a complex method). You could just do `get_selected()` and check if it's not null.

(although I haven't tested if that would work as expected, so maybe I'm wrong)

Wiwip replied:
That might be a different function. The current one is a null comparison.
`bool Tree::is_anything_selected() {
     return (selected_item != nullptr);
}`

KoBeWi replied:
You are right, I looked at wrong function 🤦‍♂️

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1923,24 +1928,27 @@ void FileSystemDock::_file_option(int p_option, const Vector<String> &p_selected
+			}
+
+			if (tree->is_anything_selected() && to_rename.path != "res://") {
+				// Edit node in Tree
```
```suggestion
				// Edit node in Tree.
```

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1923,24 +1928,27 @@ void FileSystemDock::_file_option(int p_option, const Vector<String> &p_selected
+
+				s->set_editable(tree->get_selected_column(), false);
+
+				// Set to_rename variable for callback execution
```
```suggestion
				// Set to_rename variable for callback execution.
```

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1923,24 +1928,27 @@ void FileSystemDock::_file_option(int p_option, const Vector<String> &p_selected
+				// Edit node in Tree
+				tree->grab_focus();
+				TreeItem *s = tree->get_selected();
+				s->set_editable(tree->get_selected_column(), true);
```
I think the TreeItems can be made editable when they are created (inside `_create_tree()`). Note that there are exceptions, namely `res://` and `Favorites:`.

Wiwip replied:
Yep, the "Favorites:" node is currently editable which is undesirable.
I'll check for is_anything_selected(). I based my edits on the SceneTree implementation.

I'll make a second revision tonight.

Wiwip commented:
Moved the set_editable in the _create_tree() and _update_tree(). For some reasons, the favorites creation is done within the _update_tree().
That way, the Favourite: and "res://" can't be edited.

Daylily-Zeleen committed:
Implement and expose OS::shell_show_in_file_manager()

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -3190,16 +3196,10 @@ FileSystemDock::FileSystemDock() {
-	add_child(rename_dialog);
-	rename_dialog->register_text_enter(rename_dialog_text);
-	rename_dialog->connect("confirmed", callable_mp(this, &FileSystemDock::_rename_operation_confirm));
+	current_path->set_structured_text_bidi_override(TextServer::STRUCTURED_TEXT_FILE);
+	current_path->set_h_size_flags(SIZE_EXPAND_FILL);
+	_set_current_path_text(path);
+	toolbar_hbc->add_child(current_path);
```
Looks like some copy-paste mistake. This shouldn't be here at all.

Wiwip replied:
You are totally correct. I removed the modification.

KoBeWi commented:
Wrong names are not reverted and display incorrectly:
![image](https://user-images.githubusercontent.com/2223172/206728971-ab39ae60-5281-4a32-a534-c93da2e89bfa.png)

Eoin-ONeill-Yokai committed:
Fix crash caused by invalid mix_rate assignment due to bogus
project settings.

We'll default to a sensible value in the case that a user has
somehow managed to modify the configuration file incorrectly.

Closes 69819

Wiwip commented:
I fixed the wrong names not being reverted. I considered two options:
- Simply calling update_tree() which is the simple way; or
- Restoring old_name from to_rename with some string magic which needs to handle file vs folders

i.e. `String old_name = to_rename.path.is_file ? to_rename.path.get_file() : to_rename.path.substr(0, to_rename.path.length() - 1).get_file();`
And simply go with selected_item->set_text(old_name); // approximation

My understanding is that update_tree() will be eventually called on a successful rename so it shouldn't be a performance problem to call is for a rename call.

I also noticed that accidentally renaming files such as ".tscn" or ".svg" seemed an undesired behavior.  I added a guard in the error checks since the file would be lost for the editor and hidden in the filesystem.

I haven't touched the rest of the rename_operation method since I frankly have no idea what the rest of thecode does.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -1923,24 +1943,21 @@ void FileSystemDock::_file_option(int p_option, const Vector<String> &p_selected
+					String name = to_rename.path.get_file();
+					tree->set_editor_selection(0, name.rfind("."));
+				} else {
+					String name = to_rename.path.substr(0, to_rename.path.length() - 1).get_file();
```
This can be replaced with either
```suggestion
					String name = to_rename.path.left(-1).get_file();
```
(something like that)
or
```suggestion
					String name = to_rename.path.trim_suffix("/").get_file();
```

KoBeWi commented:
> Restoring old_name from to_rename with some string magic which needs to handle file vs folders

I think this would be preferred. There are no performance considerations here, but if the tree update is not necessary it should be avoided. Although the current solution is ok too, of improving it turns out too difficult for whatever reason. You could use my suggestion above.

> I also noticed that accidentally renaming files such as ".tscn" or ".svg" seemed an undesired behavior. I added a guard in the error checks since the file would be lost for the editor and hidden in the filesystem.

Not sure what do you mean by this.

Wiwip commented:
If a file named "scene.tscn" is renamed to ".tscn" by accident or voluntarily, the rename will pass because the file length check includes the extension. On my linux machine, this also means that the file is hidden from godot as its now starting with a dot.

I made the changes, I used the suggestion as it makes sense.

However, I have to figure whether my changes have done anything because double clicks currently edit the item instead of opening them.

Wiwip commented:
The code in tree.cpp at lines 3563 is causing some weird behaviors.

Specifically, if the cell is set as "editable", a double click will automatically enable the "inline edit" mode and not send a double click signal.

```
if (rect.has_point(mpos)) {
	if (!edit_selected()) {
		emit_signal(SNAME("item_double_clicked"));
	}
} else {
	emit_signal(SNAME("item_double_clicked"));
}
```

Option 1. Optional argument to edit_selected() to bypass the selectable check and do the inline edit anyway.
Option 2. Add a flag to Tree class to allow for selection of default double click behavior. (probably the best one)

MewPurPur committed:
Use String.repeat() in more places

myaaaaaaaaa committed:
Add --quit-after <number-of-iterations>

TokageItLab committed:
Expose interpolation methods for 3D track in Animation class

dalexeev committed:
Allow negative coordinates in `AStarGrid2D`

fire committed:
Fix CSG edge case causing intersection line to hit on common edge of 2 triangles.

The previous implementation assumed that the intersection entered or exited a
shape when it hit right on the common edge of 2 triangles. However, there is
also a case where it just "skirts" the other shape on the outside.

To fix this, we added code to check the intersection distance and if the
normals of the faces are pointed in the same direction as the intersection or
not (e.g. inner product > 0). This handles the case where the intersection
line hits the common edge of 2 triangles and skirts the other shape on the
outside.

Extended code to cover a third case.

Fixes #58637.

Co-authored-by: OldBelge <StevenGeens@users.noreply.github.com>

timothyqiu committed:
Fix TouchScreenButton not redrawn when texture changes

JohanAR committed:
Add SPECULAR_AMOUNT spatial light shader built-in

Light3D has a light_specular property which is used to set the
intensity of specular contributed by this light source, but it was
previously only used by the default material light shader, and not
possible to use in a custom light() shader.

SilicDev committed:
Reimplement String.erase

bruvzg committed:
[OS] Implement and expose to scripting APIs `get_memory_info` method instead of old `get_free_static_memory`.

m4gr3d committed:
Make tab's close button responsive to touch taps

m4gr3d committed:
Make `EditorPropertyLayersGrid` responsive to touch taps

anvilfolk committed:
Fix @export variable tooltips not showing up in Inspector when there is
an inner class

mashumafi committed:
Wait for navigation baking to finish before destruction

timothyqiu committed:
Make create folder popup support nested folders

adamscott committed:
Fix GDScript LSP variable rename

paddy-exe committed:
Fix Rotation Issue with NODE_POSITION_VIEW Visual Shader Node

spanzeri committed:
Fix editor spin slider remaining editable if set read_only during an edit and fix related animation player crash

manueldun committed:
Fix auto-indentation in typed arrays, comments, and after colon

Now the editor won't add indentation when pressing enter, is
declaring typed variables and there is a colon in the comment

example:
var a:=0#:[press enter]
no indentation

Proggle committed:
TileMap: Add method to fetch the layer for a given body

vnen committed:
Add support for static variables in GDScript

Which allows editable data associated with a particular class instead of
the instance. Scripts with static variables are kept in memory
indefinitely unless the `@static_unload` annotation is used or the
`static_unload()` method is called on the GDScript.

If the custom function `_static_init()` exists it will be called when
the class is loaded, after the static variables are set.

TheSecondReal0 committed:
Add option for ButtonGroups to be unpressed

Add an option for ButtonGroups to be unpressed

Apply suggestions from code review

Co-Authored-By: Tomek <kobewi4e@gmail.com>
Co-Authored-By: Yuri Rubinsky <chaosus89@gmail.com>
Co-Authored-By: kleonc <9283098+kleonc@users.noreply.github.com>

smix8 committed:
Split Node3DGizmos into dedicated files

Splits Node3DGizmos into dedicated files.

akien-mga committed:
Merge pull request #76263 from akien-mga/ci-gotta-go-fast-with-style

CI: Speed up static checks by checking only changed files

akien-mga committed:
i18n: Sync translations with Weblate

(cherry picked from commit 90b4ca2e506c6a0d432770d59f62ae7eea565ad0)

RedworkDE committed:
C#: Do not use DocCache when generating glue code

JohanAR committed:
Add LIGHT_IS_DIRECTIONAL built-in for spatial shaders

akien-mga committed:
CI: Fixup failure to get changed files on merge

The checkout might be too shallow so the before commit isn't available.
The logic was already written to take this into account (it then generates
an empty 'changed.txt' which falls back to testing everything), but the
error code would still force terminate the job.

Hopefully we can find a way to make the logic work for merge events too in
the future, but for now this is a quick fix.

akien-mga committed:
Merge pull request #76065 from dalexeev/gds-check-tool-and-icon

GDScript: Add some checks for `@tool` and `@icon`

akien-mga committed:
Merge pull request #76269 from KoBeWi/let's_wrap_this_up

Add shortcut for quick-toggling word wrap

akien-mga committed:
Merge pull request #76286 from MystMagus/gdscript-mixed-tabs-and-spaces-fix

Fix for mixed tabs and spaces in gdscript

akien-mga committed:
Merge pull request #76288 from RedworkDE/net-glue-no-doccache

C#: Do not use DocCache when generating glue code

akien-mga committed:
Merge pull request #76296 from akien-mga/ci-static-fixup-merge

CI: Fixup failure to get changed files on merge

akien-mga committed:
CI: Fix checks fallback logic when changed files listing is missing

The file would be one byte (newline) so the check with '-s' failed.

KoBeWi committed:
Fix EditorHelp failing to load

Calinou committed:
Document custom mouse cursors larger than 128×128 potentially having issues

Calinou committed:
Add support for VRAM-compressed custom mouse cursor images

No memory is actually saved when using this, so lossless modes are
still recommended as they look better and load faster.

aaronfranke committed:
Expose `determinant` in Transform2D, rename internal method

pkowal1982 committed:
Fix lerp error message

RandomShaper committed:
Fix issues with Vulkan layout transitions

anvilfolk committed:
Improve and fix GDScript documentation generation & behavior

Removes documentation generation (docgen) from the GDScript compiler to
its own file. Adds support for GDScript enums and signal parameters and
quite a few other assorted fixes and improvements.

akien-mga committed:
Merge pull request #76306 from KoBeWi/help_is_dead,_long_live_the_help!

Fix EditorHelp failing to load

akien-mga committed:
Merge pull request #76282 from smix8/node_3d_gizmo_split_4.x

Split Node3DGizmos into dedicated files

akien-mga committed:
Merge pull request #76290 from JohanAR/light_is_directional_spatial

Add LIGHT_IS_DIRECTIONAL built-in for spatial shaders

akien-mga committed:
Merge pull request #76298 from akien-mga/ci-static-fix-fallback

CI: Fix checks fallback logic when changed files listing is missing

akien-mga committed:
Merge pull request #76308 from Calinou/doc-custom-mouse-cursor

Document custom mouse cursors larger than 128×128 potentially having issues

akien-mga committed:
Merge pull request #76314 from pkowal1982/lerp_error_message

Fix lerp error message

Commit:
Downgrade android gradle plugin to version 7.2.1.

Version 7.3.0 changes the build layout which causes updates to the generated shared libraries to not be picked up.

clayjohn committed:
Validate renderer selection in project manager and change default renderer editor setting to expose an enum to users

akien-mga committed:
Merge pull request #75960 from ShadyChibinko/tab_metadata

Add tab Metadata to TabBar & TabContainer

akien-mga committed:
Merge pull request #76309 from Calinou/custom-mouse-cursor-error-compressed

Add support for VRAM-compressed custom mouse cursor images

akien-mga committed:
Merge pull request #76325 from m4gr3d/downgrade_agp_main

Downgrade android gradle plugin to version 7.2.1.

akien-mga committed:
Merge pull request #76331 from clayjohn/PM-crash

Validate renderer selection in project manager and change default renderer editor setting to expose an enum to users

reduz committed:
Fix thread IDs.

On Linux, thread IDs were not properly assigned with the current approach.
The line:
`std::thread new_thread(&Thread::callback, _thread_id_hash(thread.get_id()), p_settings, p_callback, p_user);`
does not work because the thread ID is not assigned until the thread starts.

This PR changes the behavior to use manually generated thread IDs. Additionally, if a thread is (or may have been created) outside Godot, the method `Thread::attach_external_thread` was added.

MewPurPur committed:
Add missing PhysicalBone2D icon

AThousandShips committed:
Add information about how `Engine.time_scale` affects Timers

timothyqiu committed:
Translate undo/redo messages in settings dialogs

timothyqiu committed:
Fix layout list not resizing in editor layout dialog

MewPurPur committed:
Add an example for how to use AudioStreamGenerator

OverloadedOrama committed:
Add more signals to `TabContainer` that `TabBar` already has

KoBeWi committed:
Fix typed array export... again

Commit:
Merge pull request #75703 from m4gr3d/fix_tab_close_button_main

Fix UI responsiveness to touch taps

bruvzg committed:
[macOS] Bump min. version to 10.13, and remove deprecated code.

Rindbee committed:
Fix `line_spacing` in code editor will not take effect immediately on change

Previously, in CodeTextEditor, `line_spacing` was forgotten to update
when the editor setting property changed.

bitsawer committed:
Optimize HashMap size

sodaJar committed:
Fix wrong file on function call error

akien-mga committed:
Merge pull request #74632 from davthedev/tabs-hover

Add theming support for hovered tabs

akien-mga committed:
Merge pull request #75016 from timothyqiu/watch-out

Fix TouchScreenButton not redrawn when texture changes

akien-mga committed:
Merge pull request #75605 from anvilfolk/type-base

Make GDScript type not found errors more informative.

akien-mga committed:
Merge pull request #75640 from bruvzg/mem_info

[OS] Implement and expose to scripting APIs `get_memory_info` method instead of old `get_free_static_memory`.

akien-mga committed:
Merge pull request #75759 from TokageItLab/reimplement-grouped-statemachine

Rework for nested `AnimationNodeStateMachine`

akien-mga committed:
Merge pull request #76095 from adamscott/fix-rename-lsp-variables

[LSP] Fix GDScript inner method variable rename

akien-mga committed:
Merge pull request #76122 from spanzeri/fix_animation_snap_slider_crash

Fix editor spin slider remaining editable if set read_only during and edit and fix related animation player crash

akien-mga committed:
Merge pull request #76145 from manueldun/usability-auto-indent

Fix auto-indentation in typed arrays, comments, and after colon

akien-mga committed:
Merge pull request #76279 from TheSecondReal0/buttongroup-unpress

Add an option for ButtonGroups to be unpressed

akien-mga committed:
Merge pull request #76311 from aaronfranke/t2d-basis-det

Expose `determinant` in Transform2D, rename internal method

akien-mga committed:
Merge pull request #76352 from MewPurPur/add-physical-bone-icon

Add missing PhysicalBone2D icon

akien-mga committed:
Merge pull request #76363 from timothyqiu/whats-undo-redo

Translate undo/redo messages in settings dialogs

akien-mga committed:
Merge pull request #76364 from timothyqiu/expansion

Fix layout list not resizing in editor layout dialog

akien-mga committed:
Merge pull request #76374 from OverloadedOrama/more-tab-container-signals

Add missing `TabBar` signals to `TabContainer`

akien-mga committed:
Merge pull request #76394 from bruvzg/macos_1013

[macOS] Bump min. version to 10.13, and remove deprecated code.

akien-mga committed:
Merge pull request #76396 from Rindbee/fix_line_spacing_not_update_in_code_editor

Fix `line_spacing` in code editor will not take effect immediately on change

akien-mga committed:
Merge pull request #76403 from bitsawer/reduce_hashmap_size

Optimize HashMap size

YuriSizov committed:
Create an "Editor-only" section in the online class reference

This helps to find such classes without digging
through the rest of the class reference.
Editor-only classes can still be found under
your normal "Node" and "Resource" types.

This also fixes a typo and a missed case from the recent platform docs PR.

timothyqiu committed:
Fix inconsistent file dialog settings usage

* Project Manager ignores display mode settings.
* EditorFileDialog's default display mode and show hidden files settings
  are not updated when corresponding editor settings change.

akien-mga committed:
Merge pull request #69266 from aaronfranke/gltf-physics

Implement physics support in the GLTF module

akien-mga committed:
Merge pull request #69698 from Daylily-Zeleen/daylily-zeleen/show_in_explorer

Implement and expose OS::shell_show_in_file_manager()

akien-mga committed:
Merge pull request #73656 from TokageItLab/expose-3d-tracks

Expose interpolation methods for 3D track in `Animation` class

akien-mga committed:
Merge pull request #74101 from RandomShaper/fix_gds_obj_temps

Fix edge cases of object lifetime when signals involved

akien-mga committed:
Merge pull request #75429 from MewPurPur/Literally-cant-have-a-functioning-level-editor-because-of-this-please-helppp

Fix PackedScenes created with PackedScene.pack() have their exported nodes become null

akien-mga committed:
Merge pull request #76105 from anvilfolk/tooltip

Fix @export variable tooltips not showing up in Inspector in scripts with inner classes

akien-mga committed:
Merge pull request #76084 from timothyqiu/nested-folders

Make create folder popup support nested folders

akien-mga committed:
Merge pull request #76315 from RandomShaper/vk_fix_barriers

Fix issues with Vulkan layout transitions

akien-mga committed:
Merge pull request #76404 from sodaJar/gdrep

Fix wrong file name on function call error

akien-mga committed:
Merge pull request #76410 from YuriSizov/docs-highligh-editor-only

Create an "Editor-only" section in the online class reference

akien-mga committed:
Merge pull request #76421 from timothyqiu/editor-dialog

Fix inconsistent file dialog settings usage

bruvzg committed:
[macOS] Implement OS::shell_show_in_file_manager()

YuriSizov committed:
Unexpose AnimationTrackEditPlugin as not implemented

RandomShaper committed:
Improve reliability of editor docs cache

akien-mga committed:
Merge pull request #76413 from YuriSizov/editor-unimation-untrack-unedit-unplugin

Unexpose `AnimationTrackEditPlugin` as not implemented

akien-mga committed:
Merge pull request #76345 from reduz/fix-thread-ids

Fix thread IDs.

bitsawer committed:
Fix ShaderMaterial uniform type changes

RandomShaper committed:
Fix unsupported sampler filter used for voxel GI

akien-mga committed:
Merge pull request #76037 from theromis/master

iOS splash screen rotation fix

akien-mga committed:
Merge pull request #76168 from MewPurPur/justify-gradient-reverse-not-reversing

Document why Gradient.reverse() doesn't reverse constant gradients

akien-mga committed:
Merge pull request #76199 from needleful/incremental_search_property

Add allow_search property to ItemList and Tree to control incremental search

akien-mga committed:
Merge pull request #76246 from Proggle/tilemap-rid-layer-report

TileMap: Add method to fetch the layer for a given body

akien-mga committed:
Merge pull request #76428 from bruvzg/finder_open

[macOS] Implement OS::shell_show_in_file_manager()

akien-mga committed:
Merge pull request #76431 from RandomShaper/fix_doc_help_cache

Improve reliability of editor docs cache

akien-mga committed:
Merge pull request #69318 from Sauermann/fix-refresh-gui-events

Fix scene reload crash related to mouse cursor update

akien-mga committed:
Merge pull request #70234 from Rindbee/fix-uid-lost

Fix the uid field of the tscn/res file is lost when the external dependency is updated

akien-mga committed:
Merge pull request #75447 from bruvzg/brotli_packedarray

Expose brotli decompression to the scripting API.

akien-mga committed:
Merge pull request #75885 from AThousandShips/compound_fix

[GDScript] Fix incorrect compound assignment

dalexeev committed:
Add support for multiline strings to `Tree`

akien-mga committed:
Merge pull request #58598 from KoBeWi/shall_I_delete_ur_animations

Prompt to confirm anim track delete on node delete

akien-mga committed:
Merge pull request #61714 from dalexeev/tree-multiline-items

Add support for multiline cells to `Tree`

akien-mga committed:
Merge pull request #72421 from myaaaaaaaaa/signal-hashmap

Store Object signals in a HashMap rather than a VMap

winston-yallow committed:
Document automatic center mode using shape origins

m4gr3d committed:
Fix issue with resizing the display on Android when using the compatibility renderer.

RandomShaper committed:
Fix breakages of volumetric fog on voxel GI changes

RandomShaper committed:
Add peace-of-mind checks to API hash caching

akien-mga committed:
Merge pull request #69833 from Eoin-ONeill-Yokai/audio_invalid_mixrate

Fix crash caused by invalid mix_rate assignment due to bogus project settings.

akien-mga committed:
Merge pull request #73617 from myaaaaaaaaa/quit-after

Add `--quit-after <number-of-iterations>`

akien-mga committed:
Merge pull request #74106 from dalexeev/astar-grid-2d-negative-coords

Allow negative coordinates in `AStarGrid2D`

akien-mga committed:
Merge pull request #76359 from AThousandShips/timer_doc

Add information about how `Engine.time_scale` affects Timers

akien-mga committed:
Merge pull request #76455 from winston-yallow/document-automatic-center-of-mass

Document that the automatic center of mass mode uses the individuals shape origins

akien-mga committed:
Merge pull request #76464 from m4gr3d/fix_gl_window_resize_main

Fix issue with resizing the display on Android when using the compatibility renderer

akien-mga committed:
Merge pull request #76467 from RandomShaper/doc_cache_peace_of_mind

Add peace-of-mind checks to API hash caching

akien-mga committed:
Merge pull request #74980 from YuriSizov/build-err-without-required-editor-modules

Err when trying to build the editor without its required modules

MewPurPur committed:
Provide an example on how to use ImmediateMesh

akien-mga committed:
Merge pull request #76026 from YuriSizov/editor-running-up-that-gui

Extract editor run toolbar into its own component

akien-mga committed:
Fix various typos with codespell

Also includes the grammar fix from #76206.

Co-authored-by: Peter Anderson <BWPanda@users.noreply.github.com>

akien-mga committed:
Merge pull request #76470 from MewPurPur/add-immediate-mesh-example

Provide an example on how to use ImmediateMesh

akien-mga committed:
Merge pull request #76378 from KoBeWi/fix_typed_array_export_2-fix_harder

Fix typed array export... again

bitsawer committed:
Fix Shader Preprocessor line numbering when disabled

vnen committed:
GDScript: Don't fail when freed object is return

This is check is a bit too eager. The user should be able to handle the
return value even if it's a freed object.

Damdahlf committed:
Updated GDScript example to use clear_surfaces() and C Sharp example to use ClearSurfaces() instead of non-existent surface_remove() and SurfaceRemove()

vnen committed:
GDScript: Use default path in test generator if not provided

Since it's almost always used with the same path, assume it is running
from the Godot source root.

YuriSizov committed:
Merge pull request #72095 from anvilfolk/gd-docs

Improve GDScript documentation generation & behavior

YuriSizov committed:
Validate code tags in documentation for potential params

Adds a check to make_rst to look for matches
between the text inside of the [code][/code] tag
and known param identifiers.
Fixes most of what was revealed.

KoBeWi committed:
Improve reliability of 2D shape editor redrawing

akien-mga committed:
Merge pull request #69053 from KoBeWi/grabbercentrism

Add center_grabber theme property to Slider

akien-mga committed:
Merge pull request #76437 from RandomShaper/fix_vol_fog_voxel_gi

Fix voxel GI issues

akien-mga committed:
Merge pull request #76479 from bitsawer/fix_preprocessor_line_number

Fix Shader Preprocessor line numbering when disabled

akien-mga committed:
Merge pull request #76481 from vnen/gdscript-test-gen-default-path

GDScript: Use default path in test generator if not provided

akien-mga committed:
Merge pull request #76482 from damdalf/class_ref_documentation_MeshDataTool_code_examples

Corrected the GDScript and C# examples for the MeshDataTool class reference documentation

akien-mga committed:
Merge pull request #76488 from YuriSizov/docs-validate-code-params

Validate code tags in documentation for potential params

akien-mga committed:
Merge pull request #76492 from KoBeWi/shaped_up_nicely

Improve reliability of 2D shape editor redrawing

akien-mga committed:
Merge pull request #74053 from vnen/gdscript-fix-disassembler-lambda-crash

GDScript: Fix wrong increment for disassembly of lambda

akien-mga committed:
Merge pull request #76483 from vnen/gdscript-dont-fail-returning-freed-object

GDScript: Don't fail when freed object is return

lyuma committed:
import: Fix Silhouette used incorrect index.

Fixes bind pose mistake from using i (mesh skin index) instead of bone_idx (skeleton bone index).
Fixes #76448

RandomShaper committed:
Fix formatting function not used when printing run arguments

akien-mga committed:
Merge pull request #50671 from winston-yallow/improve-editor-spin-slider

Add editor setting for spin slider sensibility

akien-mga committed:
Merge pull request #74771 from fire/csg-fixes

Fixes to CSG robustness

akien-mga committed:
Merge pull request #76264 from vnen/gdscript-static-variales

Add support for static variables in GDScript

akien-mga committed:
Merge pull request #76499 from lyuma/fix_silhouette_fixer

import: Fix Silhouette used incorrect index.

akien-mga committed:
Merge pull request #76503 from RandomShaper/fix_args_print

Fix formatting function not used when printing run arguments

bruvzg committed:
[iOS] Fix loading of GDExtension dylibs auto converted to framework.

spanzeri committed:
Fix wait for thread not started

YuriSizov committed:
Avoid accessing the theme too early in EditorRunBar

clayjohn committed:
Use DXT1 when compressing PNGs with RGB format

This results in much smaller file sizes with the same quality

akien-mga committed:
Merge pull request #76511 from spanzeri/fix-thread-wait

Fix wait for thread not started

akien-mga committed:
Merge pull request #76438 from bitsawer/fix_shader_uniform_typing

Fix ShaderMaterial uniform type changes

akien-mga committed:
Merge pull request #76486 from Lasuch69/add-noisetexture3d-no-mipmaps

Add NoiseTexture3D

akien-mga committed:
Merge pull request #76510 from bruvzg/ios_conv_lib_load

[iOS] Fix loading of GDExtension dylibs auto converted to framework.

akien-mga committed:
Merge pull request #76513 from YuriSizov/editor-dont-run-too-fast

Avoid accessing the theme too early in EditorRunBar

akien-mga committed:
Merge pull request #76516 from clayjohn/DXT1

Use DXT1 when compressing PNGs with RGB format

justinwash committed:
Fix infinite loop in Build2DFaces::_find_edge_intersections

Tefatika committed:
Command Palette search now also uses original English command names

Both localized and non localized names will be used while filtering

The highest score between the two will be picked when determining
the entries order

bitsawer committed:
Revert "Fix ShaderMaterial uniform type changes"

This reverts commit e55c6bfb7b850b8dc9fac565ab1ce56918ea4ec7.

akien-mga committed:
GDScript: Fix template_debug compilation with TOOLS/DEBUG macro mismatch

akien-mga committed:
Merge pull request #76523 from Tefatika/feature/tefatika/76454-command-palette-use-key-name

Command Palette search now also uses original English command names

akien-mga committed:
Merge pull request #76535 from bitsawer/revert_uniform_type_check

Revert "Fix ShaderMaterial uniform type changes"

shendo committed:
[Android] Fix dynamic Variant params stack constructions in JNI callbacks

Emitting signals with params from Android plugins could crash due to
object assignment with uninitialised mem.  Instead, use 'memnew_placement'
to construct into stack addresses.  Make similar JNI callbacks consistent.

Fixes #75754.

RandomShaper committed:
Fix additional cases of breakage of rendering effects

anvilfolk committed:
Fix wrongly removing prefix when generating in-editor documentation

Co-authored-by: Danil Alexeev <danil@alexeev.xyz>

KoBeWi committed:
Fix CollisionShape2D editor crash

akien-mga committed:
Merge pull request #76539 from anvilfolk/gddocsreg

Fix wrongly removing prefix when generating in-editor documentation

akien-mga committed:
Merge pull request #76546 from KoBeWi/prevent_shape_editor_from_killing_Godot

Fix CollisionShape2D editor crash

dalexeev committed:
GDScript: Reorganize and unify warnings

akien-mga committed:
Merge pull request #75014 from vonagam/assert-literal-false

GDScript: Allow usage of literal false in assert without a warning

akien-mga committed:
Merge pull request #75732 from vonagam/fix-duplicate-condition

GDScript: Fix unnecessary duplication of pointed out lines

akien-mga committed:
Merge pull request #76412 from dalexeev/gds-reorganize-warnings

GDScript: Reorganize and unify warnings

akien-mga committed:
Merge pull request #76521 from justinwash/csg-infinite-loop-fix

Fix infinite loop in Build2DFaces::_find_edge_intersections

Tetane committed:
Use a SubViewport for CanvasItem inspector preview

clayjohn committed:
Fix compatibility breakage from adding NoiseTexture3D

Also optimize some of the Noise methods

Tefatika committed:
Fixed error messages when setting all_tab_in_front of TabContainer

Some signal handlers weren't unregistered when removing the wrapped
child TabBar, which resulted in errors printed when the TabBar was
removed and added again when the flag was updated

Calinou committed:
Document high performance cost of turbulence in ParticleProcessMaterial

- Mention that GPUParticlesAttractorVectorField3D can be used as an
  alternative to turbulence in 3D.

MewPurPur committed:
Add documentation for TubeTrailMesh and RibbonTrailMesh

KoBeWi commented on scene/gui/tree.cpp:
```diff
@@ -3834,6 +3847,10 @@ bool Tree::is_editing() {
 }
 
+void Tree::set_editor_selection(int start, int end) {
+	text_editor->select(start, end);
```
`text_editor` is now multiline, single-line editor is now under `line_editor`. This function needs to be updated.
Since single-line/multi-line depends on cell mode, maybe selection rows could be optional arguments.

KoBeWi commented on scene/gui/tree.h:
```diff
@@ -700,6 +708,7 @@ class Tree : public Control {
 	Rect2 get_item_rect(TreeItem *p_item, int p_column = -1, int p_button = -1) const;
 	bool edit_selected();
 	bool is_editing();
+	void set_editor_selection(int start, int end);
```
```suggestion
	void set_editor_selection(int p_start, int p_end);
```

KoBeWi commented on scene/gui/tree.cpp:
```diff
@@ -3561,8 +3569,13 @@ void Tree::gui_input(const Ref<InputEvent> &p_event) {
+									emit_signal(SNAME("item_double_clicked"));
+									break;
+								case INLINE_EDIT:
+									edit_selected();
```
Probably should fallback to emit signal if the item is uneditable, to preserve previous behavior.

HolonProduction committed:
Preserve scene unique names when saving branch as scene.

KoBeWi commented:
> I fixed the wrong names not being reverted.

Seems to be still happening. Tested on folders with duplicate names.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -69,9 +69,11 @@ bool FileSystemDock::_create_tree(TreeItem *p_parent, EditorFileSystemDirectory
 
 	// Create a tree item for the subdirectory.
 	TreeItem *subdirectory_item = tree->create_item(p_parent);
+	subdirectory_item->set_editable(0, true);
```
I think this can be moved to `else` for the condition below. No reason to set something and then set back.

KoBeWi committed:
Some clarifications on screen-space coordinates

Koyper committed:
Update @GDScript.xml range docs to iterate backwards.

Commit:
Merge pull request #76640 from shendo/android_plugin_crash_on_emit_signal

[Android] Fix dynamic Variant params stack constructions in JNI callbacks

brno32 committed:
add venv and .venv to the .gitignore

Xenoparrot committed:
Fixed various typos across the docs

smix8 committed:
Fix SurfaceTool::create_from_blend_shape()

Fixes SurfaceTool::create_from_blend_shape().

clayjohn committed:
Merge pull request #76597 from Tefatika/fix/tefatika/76578-tab-container-all_tab_in_front-error

Fixed error messages when setting all_tab_in_front of TabContainer

clayjohn committed:
Merge pull request #76598 from Calinou/doc-particleprocessmaterial-turbulence-cost

Document high performance cost of turbulence in ParticleProcessMaterial

brno32 committed:
adjusting step size for fastnoise frequency hint

clayjohn committed:
Merge pull request #68033 from MewPurPur/auto-light3d-warning-updates

Show Light3D's scale warning immediately when a user edits the property.

clayjohn committed:
Use proper UV in cubemap downsampler raster

This removes bias in cubemap downsampling shader that resulted in the bottom of cubemaps being over represented

QbieShay committed:
Merge pull request #76683 from brno32/fastnoise-lite-frequency-hint

adjusting step size for fastnoise frequency hint

MewPurPur committed:
Document TextServerDummy

dsnopek committed:
Explicitly mark inherited export mode when making a dedicated server export

clayjohn committed:
Merge pull request #76609 from HolonProduction/save-branch-as-scene-unique-name

Preserve scene unique names when saving branch as scene.

mhilbrunner committed:
Merge pull request #76555 from Tetane/master

Fix 2D shader preview draws over uniform

mhilbrunner committed:
Merge pull request #76669 from smix8/fix_create_from_blend_shape_4.x

Fix SurfaceTool::create_from_blend_shape()

mhilbrunner committed:
Merge pull request #76692 from clayjohn/Mobile-sky-dark

Use proper UV in cubemap downsampler raster (Fixes reflections in mobile renderer)

mhilbrunner committed:
Merge pull request #76557 from clayjohn/NoiseTexture3D

Fix compatibility breakage from adding NoiseTexture3D

mhilbrunner committed:
Merge pull request #76550 from RandomShaper/vk_gi_take_2

Fix voxel GI issues (2)

mhilbrunner committed:
Merge pull request #76646 from KoBeWi/acoording_to_screen

Some clarifications on screen-space coordinates

aaronfranke committed:
Fix small mistake in docs, "antecedents" -> "ancestors"

Commit:
Make sun energy adjustable as a float in the 3D editor's preview sun

clayjohn committed:
Merge pull request #76722 from HotHead007/master

Make sun energy adjustable as a float in the 3D editor's preview sun

spanzeri committed:
Support long path in file access on windows

Changed windows file access file to check for path length and use the \\?\ long format when needed

Commit:
Make environment energy adjustable as a float in 3D editor's preview environment

mhilbrunner committed:
Merge pull request #58518 from Jayman2000/bug-reporting-guidelines

Streamline and centralize bug reporting guides

mhilbrunner committed:
Merge pull request #76739 from spanzeri/fix-long-path-windows

mhilbrunner committed:
Merge pull request #76004 from mashumafi/wait-nav-region-thread

mhilbrunner committed:
Merge pull request #75055 from JohanAR/light_specular

mhilbrunner committed:
Merge pull request #76604 from MewPurPur/sylveon-moment-UwU

mhilbrunner committed:
Merge pull request #74883 from dalexeev/fix-thin-multiline-colors

mhilbrunner committed:
Merge pull request #75517 from BastiaanOlij/expose_render_target

mhilbrunner committed:
Merge pull request #76109 from paddy-exe/fix-node-position-view

mhilbrunner committed:
Merge pull request #75864 from KoBeWi/assassin_of_shaders

mhilbrunner committed:
Merge pull request #76650 from Koyper/gdscript_range_docs

clayjohn committed:
Merge pull request #75427 from Chaosus/shader_fix_groups_ordering

Fix shader uniform groups to prevent it sorting in alphabet order

clayjohn committed:
Merge pull request #76700 from dsnopek/dedicated-server-mark-inherited

Explicitly mark inherited export mode when making a dedicated server export

clayjohn committed:
Merge pull request #72288 from MewPurPur/use-string-repeat

Use `String.repeat()` to optimize several String methods

clayjohn committed:
Merge pull request #75510 from SilicDev/string_erase

Reimplement `String.erase()` as immutable method

novalis committed:
Explain why an image would be invalid for a texture

KoBeWi committed:
Remove erroneous signal connection

clayjohn committed:
Merge pull request #76765 from novalis/explain-why-invalid

Explain why an image would be invalid for a texture

mhilbrunner committed:
Merge pull request #76743 from HotHead007/master

mhilbrunner committed:
Merge pull request #76761 from KoBeWi/this_file_ain't_big_enough_for_both_of_us

mhilbrunner committed:
Merge pull request #76716 from brno32/venv-gitignore

mhilbrunner committed:
Merge pull request #76720 from aaronfranke/ancestortecedent

mhilbrunner committed:
Merge pull request #76699 from MewPurPur/document-text-server-dummy

mhilbrunner committed:
Merge pull request #76664 from Xenoparrot/simple-typos

mhilbrunner committed:
Merge pull request #76373 from MewPurPur/add-AudioStreamGenerator-example

Wiwip committed:
Update scene/gui/tree.h

Co-authored-by: Tomek <kobewi4e@gmail.com>

Wiwip committed:
Proposal to replace the rename editor with the line_editor

Wiwip committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	editor/filesystem_dock.cpp
#	scene/gui/tree.cpp
#	scene/gui/tree.h

Wiwip commented:
Alright, I have made changes, but also a huge mistake with git I believe.

Wiwip commented:
I created a new pull request instead

https://github.com/godotengine/godot/pull/76794
